libshaderc_spvc/src/spvc.cc

```diff
@@ -164,74 +174,370 @@ void shaderc_spvc_compiler_release(shaderc_spvc_compiler_t compiler) {
+
+        switch (mode) {
+          case spv::ExecutionModeInvocations:
+            GetOpsRecur(inst, 2, execution.invocations);
+            break;
+          case spv::ExecutionModeLocalSize:
+            GetOpsRecur(inst, 2, execution.workgroup_size.x,
+                        execution.workgroup_size.y, execution.workgroup_size.z);
+            break;
+          case spv::ExecutionModeOutputVertices:
+            GetOpsRecur(inst, 2, execution.output_vertices);
+            break;
+          default:
+            break;
+        }
+      } break;
+
+      case spv::OpCapability: {
+        spv::Capability cap;
+        GetOps(inst, cap);
+        if (cap == spv::CapabilityKernel) {
+          result->messages.append("Kernel capability not supported.");
+          return SPV_REQUESTED_TERMINATION;
+        }
+        ir.declared_capabilities.push_back(cap);
+      } break;
+
+      case spv::OpTypeVoid: {
+        uint32_t id;
+        GetOps(inst, id);
+        auto& type = set<spirv_cross::SPIRType>(inst->result_id);
+        type.basetype = spirv_cross::SPIRType::Void;
+      } break;
+
+      case spv::OpTypeFunction: {
+        uint32_t id;
+        uint32_t ret;
+        auto& func =
+            set<spirv_cross::SPIRFunctionPrototype>(inst->result_id, ret);
+        GetOps(inst, id, ret, func.parameter_types);
+      } break;
+
+      case spv::OpFunction: {
+        uint32_t res, id, control, type;
+        GetOps(inst, res, id, control, type);
+
+        if (current_function) {
+          result->messages.append(
+              "Must end a function before starting a new one!");
+          return SPV_REQUESTED_TERMINATION;
+        }
+
+        current_function =
+            &set<spirv_cross::SPIRFunction>(inst->result_id, res, type);
+      } break;
+
+      case spv::OpFunctionEnd: {
+        if (current_block) {
+          result->messages.append(
+              "Cannot end a function before ending the current block.\nLikely "
+              "cause: If this SPIR-V was created from glslang HLSL, make sure "
+              "the entry point is valid.");
+          return SPV_REQUESTED_TERMINATION;
+        }
+        current_function = nullptr;
+      } break;
+
+      case spv::OpLabel: {
+        if (!current_function) {
+          result->messages.append("Blocks cannot exist outside functions!");
+          return SPV_REQUESTED_TERMINATION;
+        }
+        uint32_t id;
+        GetOps(inst, id);
+
+        current_function->blocks.push_back(inst->result_id);
+        if (!current_function->entry_block) {
+          current_function->entry_block = inst->result_id;
+        }
+
+        if (current_block) {
+          result->messages.append(
+              "Cannot start a block before ending the current block.");
+          return SPV_REQUESTED_TERMINATION;
+        }
+
+        current_block = &set<spirv_cross::SPIRBlock>(inst->result_id);
+      } break;
+
+      case spv::OpReturn: {
+        if (!current_block) {
```

Comment:
Should we move these blocks into a method per instruction? This method is going to get really long as it is.

Reply:
sure